src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -408,16 +468,28 @@ private PingReply GetAddressAndSend(string hostNameOrAddress, int timeout, byte[
             }
         }
 
-        private async Task<PingReply> GetAddressAndSendAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions? options)
+        private async Task<PingReply> SendPingAsyncInternal<TArg>(
+            TArg getAddressArg,
+            Func<TArg, CancellationToken, ValueTask<IPAddress>> getAddress,
+            int timeout,
+            byte[] buffer,
+            PingOptions? options,
+            CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             CheckStart();
             try
             {
-                IPAddress[] addresses = await Dns.GetHostAddressesAsync(hostNameOrAddress).ConfigureAwait(false);
-                Task<PingReply> pingReplyTask = SendPingAsyncCore(addresses[0], buffer, timeout, options);
-                return await pingReplyTask.ConfigureAwait(false);
+                using CancellationTokenRegistration _ = cancellationToken.Register(static state => ((Ping)state!).SetCanceled(), this);
```

Comment:
It's unlikely to make a material difference, but it looks like this could use UnsafeRegister instead of Register.